lib/WTSI/DNAP/Warehouse/Schema/Result/EseqFlowcell.pm

```diff
@@ -356,8 +356,58 @@ __PACKAGE__->belongs_to(
 
 # You can replace this text with custom code or comments, and it will be preserved on regeneration
 
+use Readonly;
+
 our $VERSION = '0';
 
+=head1 SUBROUTINES/METHODS
+
+=head2 sample_id
+
+=head2 sample_name
+
+=head2 sample_supplier_name
+
+=head2 study_name
+
+=cut
+
+#####
+# Delegations below are provided to ensure compatibility with SeqQC viewer code.
+#
+
+Readonly my %DELEGATION_TO_SAMPLE => {
+    'sample_id'                => 'id_sample_lims',
+    'sample_name'              => 'name',
+    'sample_supplier_name'     => 'supplier_name',
+};
+
+Readonly my %DELEGATION_TO_STUDY => {
+    'study_name'                          => 'name',
+};
+
+foreach my $rel (qw(sample study)) {
+
+  my $attr = q[_] . $rel . q[_row];
+  my $del  = $rel eq 'sample' ? \%DELEGATION_TO_SAMPLE : \%DELEGATION_TO_STUDY;
+
+  has $attr => ( isa        => 'Maybe[WTSI::DNAP::Warehouse::Schema::Result::' . ucfirst $rel . ']',
+                 is         => 'ro',
+                 weak_ref   => 1,
+                 lazy_build => 1,
+                 handles    => $del,
+  );
+
+  __PACKAGE__->meta->add_method('_build_' . $attr, sub {my $r = shift; return $r->$rel;} );
+
+  foreach my $method ( keys %{$del} ) {
+    around $method => sub {
+      my ($orig, $self) = @_;
+      return $self->$attr ? $self->$orig() : undef;
```

Comment:
This is confusing me.

So if I call `$self->sample_id` the handler causes `$self->sample` to be lazy-built by calling `$r->sample`. Is `$r` self at that point, or the class?

Somehow this resolves to a DBIC join.

Reply:
Yes, `$r` is `$self`.  I think the intent was to avoid resolving the `sample` and `study` joins multiple time. The relevant sample and study rows are cached.